stdlib/public/core/FloatingPointTypes.swift.gyb

```diff
@@ -760,25 +761,45 @@ extension ${Self} {
 extension ${Self} {
 % for srcBits in allFloatBits:
 %   That = floatName[srcBits]
-%   if Self != That:
 
-%     if srcBits == 80:
+%   if srcBits == 80:
 #if !os(Windows) && (arch(i386) || arch(x86_64))
-%     end
+%   end
 
+%   if srcBits == bits:
+  /// Create an instance initialized to `value`.
+%   else:
   /// Construct an instance that approximates `other`.
+%   end
+  @_transparent
   public init(_ other: ${That}) {
-%     if srcBits > bits:
+%   if srcBits > bits:
     _value = Builtin.fptrunc_FPIEEE${srcBits}_FPIEEE${bits}(other._value)
-%     else:
+%   elif srcBits < bits:
     _value = Builtin.fpext_FPIEEE${srcBits}_FPIEEE${bits}(other._value)
-%     end
+%   else:
+    _value = other._value
+%   end
+  }
+  
+  /// Create an ${That} initialized to `value`, if `value` can be represented without rounding.
+  /// - note: This is effectively checking that `inputValue == outputValue`, meaning `NaN` inputs will always return `nil`. 
+  @inline(__always)
+  public init?(exactly other: ${That}) {
+    if other.isNaN {
+      return nil
+    }
+    self.init(other)
+%   if srcBits > bits:
+    // Infinite conversions are always OK. If not infinite, check that we can round-trip convert and get the same value.
```

Comment:
"Infinite conversions" -> "Converting the infinity value is considered value preserving.  In other cases, check that we can round-trip and get the same value."  I think this might be a better wording.